### PR TITLE
*: support record kv/pd time for a SQL statement

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/17367

also fixes https://github.com/pingcap/tidb/issues/17361 by hand

Problem Summary:

we are missing record many cost time in the current slow log, for example:

- point-get network/kv time or time for backoff when it meet error in point-get
- batch-check/rebuild index check network/kv time or time for backoff when they meet
- resolve lock network/kv time(getStatus+resolve 2 rpc) and backoff when it meets
- some request met region cache miss and request pd to get region info rpc time
- .....

### What is changed and how it works?

What's Changed, How it Works:

the question is we are hard to pass execDetail for every this low level operation and record time.]

so in this PR, we change the other way: just add an interceptor for pd/kv client and record time in context.Context.

so we can record time no matter how deep the method is if we can pass context.Context in.

### Related changes

- n/a

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - addition small memory alloc and time.Now call

### Release note 

- Support record kv/pd time for a SQL statement

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/17591)

### CODE REVIEW

File: executor/adapter.go [hunk 1]
```diff
@@ -175,6 +178,7 @@ type ExecStmt struct {
 	isPreparedStmt    bool
 	isSelectForUpdate bool
 	retryCount        uint
+	IsInMultiQueries  bool
```
Comment (tiancaiamao) [inline executor/adapter.go:181]: We do not use InMultiQueries after this refactor https://github.com/pingcap/tidb/pull/16056

File: infoschema/tables.go [hunk 2]
```diff
@@ -1578,7 +1578,7 @@ func (it *infoschemaTable) RemoveRecord(ctx sessionctx.Context, h kv.Handle, r [
 }
 
 // UpdateRecord implements table.Table UpdateRecord interface.
-func (it *infoschemaTable) UpdateRecord(ctx sessionctx.Context, h kv.Handle, oldData, newData []types.Datum, touched []bool) error {
+func (it *infoschemaTable) UpdateRecord(gctx context.Context, ctx sessionctx.Context, h kv.Handle, oldData, newData []types.Datum, touched []bool) error {
```
Comment (tiancaiamao) [inline infoschema/tables.go:1581]: Better  use  `ctx` for `context.Context` and `sctx` for `sessionctx.Context`, the former is a standard one.

File: infoschema/tables.go [hunk 3]
```diff
@@ -1705,7 +1705,7 @@ func (vt *VirtualTable) RemoveRecord(ctx sessionctx.Context, h kv.Handle, r []ty
 }
 
 // UpdateRecord implements table.Table UpdateRecord interface.
-func (vt *VirtualTable) UpdateRecord(ctx sessionctx.Context, h kv.Handle, oldData, newData []types.Datum, touched []bool) error {
+func (vt *VirtualTable) UpdateRecord(gctx context.Context, ctx sessionctx.Context, h kv.Handle, oldData, newData []types.Datum, touched []bool) error {
```
Comment (tiancaiamao) [inline infoschema/tables.go:1708]: ditto

File: session/session.go [hunk 4]
```diff
@@ -1212,6 +1211,7 @@ func (s *session) execute(ctx context.Context, sql string) (recordSets []sqlexec
 	multiQuery := len(stmtNodes) > 1
```
Comment (tiancaiamao) [inline session/session.go:1211]: This function is not used, or won't be used  in the future.
`Execute()` returning multiple record sets are deprecated.

- The normal query runs through ExecuteStmt
- Prepared query runs through ExecutePreparedStmt
- Only internal SQLs are still running to here, and I'm cleaning up the code.

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -1531,6 +1531,8 @@ const (
 	SlowLogPlanFromCache = "Plan_from_cache"
 	// SlowLogHasMoreResults is used to indicate whether this sql has more following results.
 	SlowLogHasMoreResults = "Has_more_results"
+	// SlowLogInMultiQueries is used to indicate whether this sql is executed in multi query..
```
Comment (tiancaiamao) [inline sessionctx/variable/session.go:1534]: Not  needed

File: sessionctx/variable/session.go [hunk 6]
```diff
@@ -1576,9 +1582,13 @@ type SlowQueryLogItems struct {
 	Prepared       bool
 	PlanFromCache  bool
 	HasMoreResults bool
+	InMultiQueries bool
```
Comment (tiancaiamao) [inline sessionctx/variable/session.go:1585]: Still here...

**Comment by crazycs520** [comment]
@djshow832 PTAL, It can be added into statement too.

**Comment by djshow832** [comment]
> @djshow832 PTAL, It can be added into statement too.

OK

File: table/tables/tables.go [hunk 7]
```diff
@@ -412,7 +412,7 @@ func (t *TableCommon) UpdateRecord(ctx sessionctx.Context, h kv.Handle, oldData,
 	return nil
 }
 
-func (t *TableCommon) rebuildIndices(ctx sessionctx.Context, rm kv.RetrieverMutator, h kv.Handle, touched []bool, oldData []types.Datum, newData []types.Datum) error {
+func (t *TableCommon) rebuildIndices(ctx sessionctx.Context, rm kv.RetrieverMutator, h kv.Handle, touched []bool, oldData []types.Datum, newData []types.Datum, opts ...table.CreateIdxOptFunc) error {
```
Comment (jackysp) [inline table/tables/tables.go:415]: Is it a bug fix?

File: infoschema/tables.go (hunk 2 above)
Comment (jackysp) [reply infoschema/tables.go:1596]: Please address the comment or reply it.

**Comment by jackysp** [comment]
Please resolve the conflicts.

File: server/conn.go [hunk 8]
```diff
@@ -1291,6 +1292,7 @@ func (cc *clientConn) handleQuery(ctx context.Context, sql string) (err error) {
 }
 
 func (cc *clientConn) handleStmt(ctx context.Context, stmt ast.StmtNode, lastStmt bool) error {
+	ctx = context.WithValue(ctx, execdetails.StmtExecDetailKey, &execdetails.StmtExecDetails{})
```
Comment (crazycs520) [inline server/conn.go:1295]: How about add-in `execRestrictedSQL` too? To record for internal SQL.

File: sessionctx/variable/session.go (hunk 6 above)
Comment (crazycs520) [reply sessionctx/variable/session.go:1591]: It's useless?

File: util/execdetails/execdetails.go [hunk 9]
```diff
@@ -47,6 +47,19 @@ type ExecDetails struct {
 	CommitDetail     *CommitDetails
 }
 
+type stmtExecDetailKeyType struct{}
+
+// StmtExecDetailKey used to carry StmtExecDetail info in context.Context.
+var StmtExecDetailKey = stmtExecDetailKeyType{}
+
+// StmtExecDetails contains stmt level execution detail info.
+type StmtExecDetails struct {
```
Comment (crazycs520) [inline util/execdetails/execdetails.go:56]: How about record the Backoff by backoff_type?
And record the RPC request type too?

Such as below:
```shell
# Backoff_total_time: 10 Backoff_regionmiss_time: 8 Backoff_noleader_time: 2
# TIKV_rpc_total_time: 20 TiKV_rpc_cop_time: 15 TiKV_rpc_resolveLock_time: 5
```

Comment (lysu) [reply util/execdetails/execdetails.go:56]: emmm..  I also think this is better to record fine-grained time，but it should introduce more performance cost, how about record total at this pr, and do fine-grained time if need in future? :D

File: sessionctx/variable/session.go [hunk 10]
```diff
@@ -1603,6 +1603,14 @@ const (
...
 	// SlowLogPrevStmtPrefix is the prefix of Prev_stmt in slow log file.
 	SlowLogPrevStmtPrefix = SlowLogPrevStmt + SlowLogSpaceMarkStr
+	// SlowLogKVTotal is the total time waiting for kv.
+	SlowLogKVTotal = "KV_total"
+	// SlowLogPDTotal is the total time waiting for pd.
+	SlowLogPDTotal = "PD_total"
+	// SlowLogBackoffTotal is the total time doing backoff.
+	SlowLogBackoffTotal = "Backoff_total"
+	// SlowLogRespClientTime is the total time used to write response to client.
+	SlowLogRespClientTime = "Resp_client_time"
```
Comment (zz-jason) [inline sessionctx/variable/session.go:1613]: the name is a little confusing.

Comment (lysu) [reply sessionctx/variable/session.go:1613]: how about "Write_sql_response_time" ~?

**Pull request merged**